src/metamodels/generic.py

```diff
@@ -0,0 +1,127 @@
+from collections import namedtuple
+
+
+class GenericMetamodel(IBayesDBMetamodel):
+    # TODO: Code goes here.
+    pass
+
+
+class ColumnQuery(namedtuple('_ColumnQuery', ['row', 'colno'])):
+    '''Specifies a cell in the data. row can be None to indicate "unobserved."
+    colno is zero-based.'''
+    pass
+
+
+class ColumnValue(namedtuple('_ColumnValue', ['row', 'colno', 'value'])):
+    '''Similar to ColumnQuery, but with a value assigned to that cell, in the
+    original domain of the user-supplied data.'''
+    pass
+
+
+class GenerativePopulationModel(object):
+    '''See the BayesDB paper for the semantics.'''
+
+    def simulate(self, givens, targets):
+        '''givens is a list of ColumnValue. targets is a list of
+        ColumnQuery.'''
+        raise NotImplementedError
+
+    def logpdf(self, givens, query):
+        '''givens and query are lists of ColumnValue.'''
+        raise NotImplementedError
+
+    def kl_divergence(self, measurements_a, measurements_b):
+        '''measurements_a and measurements_b are lists of ColumnQuery.'''
+        raise NotImplementedError
+
+
+class ModelState(object):
+    '''Represents a mutable model in training. Corresponds to a single BQL
+    model.'''
+
+    def serialize(self):
+        '''Serializes the state of this object to a bytes and returns it.'''
+        raise NotImplementedError
+
+    @classmethod
+    def deserialize(cls, data):
+        '''Recreates a ModelState from a bytes returned by serialize. Returns a
+        ModelState.'''
+        raise NotImplementedError
+
+    @classmethod
+    def schema_column_names(cls, schema):
+        '''schema is a list-of-lists representing a BQL schema. See
+        IBayesDBMetamodel.create_generator for the format. Returns a list of
+        unicode column names.'''
+
+    @classmethod
+    def from_records(cls, array, column_names, schema, random_seed):
+        '''Creates a new ModelState from array, a numpy array with columns
+        named by column_names. These column_names were extracted from schema by
+        an earlier call to schema_column_names with the same schema. bayeslite
+        can't implement it with a single call because it needs to know the
+        names of the columns to load into array.'''
+        raise NotImplementedError
+
+    @property
+    def iterations():
+        '''The number of Markov chain transitions completed by this state.'''
+        raise NotImplementedError
+
+    def transition_iterations(self, iterations):
+        '''Performs the specified number of Markov chain iterations. If it is
+        interrupted, no methods will ever be called on the object again.'''
+        raise NotImplementedError
+
+    def transition_seconds(self, seconds):
+        '''Performs the Markov chain iterations for the specified number of
+        (floating point) seconds. If it is interrupted, no methods will ever be
+        called on the object again.'''
+        raise NotImplementedError
+
```

Comment:
Consider merging `transition_seconds` and `transition_iterations` into `transtion(iters=None, seconds=None)`. If only `iters` (viz `seconds`) is specified, then that amount of anlaysis will be conducted. If both are specified, then analysis shall terminate at whichever comes first. I have found this functionality extremely useful in gpmcc, in particular the "timing-out" parameter `seconds` when coupled with `iterations` has saved me when:
- I underestimated the amount of time it would take to run `N` iterations, but interrupting analysis will kill all my progress thus far.
- A small set of "straggler" models in the ensemble are taking order of magnitudes longer than the other models to complete `N` iterations, holding back my workflow (this issue could theoretically be circumvented by continuous or background analysis, but until then the solution above works well).